Add a shape inference tool

GD06 opened the pull request:
Summary:
Add a shape inference tool that helps to infer each node shape of a given graph module.
1. Given a fx graph, and an example of an input(don't need to be an accurate input that can be forward, but should have valid dims and data structures), `infer shape` creates an input of symbolic shape
2. Shape prop this symbolic input can catch runtime or value exceptions.
3. These errors are constraints for symbol values, and the constraint solver `infer symbolic values` helps us figure out specific values for each symbol.
4. Finally, we run the shape propagation based on input tensor to get tensor shapes for all nodes in the FX traced module.

Test Plan:
### 1. Test `infer symbol values`
Command:
```
buck2 test mode/opt //caffe2/test:fx_experimental -- test_infer_symbol_values
```

### 2. Test `infer shape`
Command:
```
buck2 test mode/opt //caffe2/test:fx_experimental -- test_infer_symbol_values
```
Inferred shape result like: P897560514

Differential Revision: D53593702

yf225 commented on test/test_fx_experimental.py:
```diff
@@ -28,6 +30,11 @@
+from torch.fx.experimental.shape_inference.infer_shape import infer_shape
+from torch.fx.experimental.shape_inference.infer_symbol_values import infer_symbol_values
+from dper_lib.fx_tracing.graph_module import DperGraphModule
+from dper_lib.fx_tracing.tracer import DperTracer
```
we likely don't have these two in OSS

GD06 replied:
Fixed.

yf225 commented on test/test_fx_experimental.py:
```diff
@@ -7,9 +7,11 @@
 import sys
```
ideally we can put this test file in a sub folder like `test/fx/`

GD06 replied:
Moved.

GD06 commented:
Hi @yf225 , we have addressed the review comments, could you please take a second look at this PR?

yf225 commented on torch/fx/experimental/shape_inference/infer_shape.py:
```diff
@@ -0,0 +1,98 @@
+
+
+"""
+This is the function that runs shape inference will modify the input graph module so that shapes are annotated.
```
maybe it should be:
```suggestion
This is the function that runs shape inference. It will modify the input graph module so that shapes are annotated.
```
and we can move this comment to be within the `infer_shape` function.

GD06 committed:
Add a shape inference tool (#120097)

Summary:
Add a shape inference tool that helps to infer each node shape of a given graph module.
1. Given a fx graph, and an example of an input(don't need to be an accurate input that can be forward, but should have valid dims and data structures), `infer shape` creates an input of symbolic shape
2. Shape prop this symbolic input can catch runtime or value exceptions.
3. These errors are constraints for symbol values, and the constraint solver `infer symbolic values` helps us figure out specific values for each symbol.
4. Finally, we run the shape propagation based on input tensor to get tensor shapes for all nodes in the FX traced module.


Test Plan:
### 1. Test `infer symbol values`
Command:
```
buck2 test mode/opt //caffe2/test:fx_experimental -- test_infer_symbol_values
```

### 2. Test `infer shape`
Command:
```
buck2 test mode/opt //caffe2/test:fx_experimental -- test_infer_symbol_values
```
Inferred shape result like: P897560514

Differential Revision: D53593702

Pulled By: GD06

GD06 commented:
@pytorchmergebot

GD06 commented:
@pytorchmergebot  merge
